klippy/extras/palette2.py

```diff
@@ -0,0 +1,610 @@
+        except SerialException:
+            gcmd.respond_info("Unable to connect to the Palette 2")
+            return
+
+        reactor = self.printer.get_reactor()
+        if self.read_timer is None:
+            self.read_timer = reactor.register_timer(
+                self._run_Read, reactor.NOW)
+        if self.write_timer is None:
+            self.write_timer = reactor.register_timer(
+                self._run_Write, reactor.NOW)
+
+        # Tell the device we're alive
+        lastHeartbeatSend = time.time()
+        self.write_queue.put("\n")
+        self.write_queue.put(COMMAND_HEARTBEAT)
+        self.write_queue.put(COMMAND_FIRMWARE)
+
+    cmd_Disconnect_Help = ("Disconnect from the Palette 2")
+
+    def cmd_Disconnect(self, gmcd=None):
+        logging.info("Disconnecting from Palette 2")
+        if self.serial:
+            self.serial.close()
+            self.serial = None
+
+        reactor = self.printer.get_reactor()
+        reactor.unregister_timer(self.read_timer)
+        self.read_timer = None
+        reactor.unregister_timer(self.write_timer)
+        self.write_timer = None
+
+        with self.write_queue.mutex:
+            self.write_queue.queue.clear()
+        self.heartbeat = None
+        self.is_printing = False
+
+    cmd_Clear_Help = ("Clear the input and output of the Palette 2")
+
+    def cmd_Clear(self, gcmd):
+        logging.info("Clearing Palette 2 input and output")
+        if self._check_P2(gcmd):
+            for l in COMMAND_CLEAR:
+                self.write_queue.put(l)
+
+    cmd_Cut_Help = ("Cut the outgoing filament")
+
+    def cmd_Cut(self, gcmd):
+        logging.info("Cutting outgoing filament in Palette 2")
+        if self._check_P2(gcmd):
+            self.write_queue.put(COMMAND_CUT)
+
+    cmd_Smart_Load_Help = ("Automatically load filament through the extruder")
+
+    def cmd_Smart_Load(self, gcmd):
+        if self._check_P2(gcmd):
+            if not self.is_loading:
+                gcmd.respond_info(
+                    "Cannot auto load when the Palette 2 is not ready")
+                return
+            self.cmd_P2_O102(params=None)
+
+    def cmd_OmegaDefault(self, gcmd):
+        logging.debug("Omega Code: %s" % (gcmd.get_command()))
+        if self._check_P2(gcmd):
+            self.write_queue.put(gcmd.get_commandline())
+
+    cmd_O1_help = (
+        "Initialize the print, and check connection with the Palette 2")
+
+    def cmd_O1(self, gcmd):
+        logging.info("Initializing print with Pallete 2")
+        if not self._check_P2(gcmd):
+            raise self.printer.command_error(
+                    "Cannot initialize print, palette 2 is not connected")
+
+        startTs = time.time()
+        while self.heartbeat is None and self.heartbeat < (
+                time.time() -
+                HEARTBEAT_TIMEOUT) and startTs > (
+                time.time() -
+                HEARTBEAT_TIMEOUT):
+            time.sleep(1)
```

Comment:
It's not valid to call `time.sleep()` in klippy - it'll hang the entire print process.  Use `reactor.pause()` instead.

Reply:
Noted, I'll make the change